### PR TITLE
Add support for broadcasting to all panes in a tab

### LINKED ISSUE #2634: Support broadcast input?
# Description of the new feature/enhancement

Sometimes i need to open many panes in a tab with splitHorizontal/splitVertical hotkeys. And then broadcast input to all the panes in current tab, or all tabs.

Is it possible to support broadcast input, like iterm2?

image

### PR DESCRIPTION
**Pull request by zadjii-msft**
Resurrection of #9222. 
Spec draft in #9365.

Consensus from community feedback is that the whole of that spec is _nice to have_, but what REALLY matters is just broadcasting to all the panes in a tab. So, in the interest of best serving our community, I'm pushing this out as the initial implementation, before we figure out the rest of design. Regardless of how we choose to implement the rest of the features detailed in the spec, the UX for this part of the feature remains the same. 

This PR adds a new action: `toggleBroadcastInput`. Performing this action starts broadcasting to all panes in this tab. Keystrokes in one pane will be sent to all panes in the tab. 

An icon in the tab is used to indicate when this mode is active. Furthermore, the borders of all panes will be highlighted with `SystemAccentColorDark2`/`SystemAccentColorLight2` (depending on the theme), to indicate they're also active.

* [x] Closes #2634. 
  - (we should lick a reserved thread for follow-ups)

Co-authored-by: Don-Vito khvitaly@gmail.com

### CODE REVIEW

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 1]
```diff
@@ -517,6 +533,19 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
                                       const ControlKeyStates modifiers,
                                       const bool keyDown)
+    {
+        auto keySentArgs = winrt::make<KeySentEventArgs>(vkey, scanCode, modifiers, keyDown);
+        _KeySentHandlers(*this, keySentArgs);
+        return BroadcastKeyEvent(vkey, scanCode, modifiers, keyDown);
+    }
+
+    // Call this method to bypass raising a KeySent. TrySendKeyEvent should be
+    // called when teh event originated in this control. BroadcastKeyEvent is
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:543]:
```suggestion
    // called when the event originated in this control. BroadcastKeyEvent is
```

(scrub)

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 2]
```diff
@@ -1190,6 +1203,21 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         }
     }
 
+    bool TermControl::BroadcastKeyEvent(const WORD vkey,
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp:1206]: I'd caution on this naming _in all layers_. This is "a key event was broadcasted directly to me," which really is just closer to "`RawWriteKeyEvent`". Right now it seems like this will _cause_ a broadcast (and the immediate next questions are "to whom? from whom?"

Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp:1206]: Additionally, is there a way to converge this with `IDirectKeyListener`'s manual key handler?

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 3]
```diff
@@ -517,6 +533,19 @@ namespace winrt::Microsoft::Terminal::Control::implementation
                                       const WORD scanCode,
                                       const ControlKeyStates modifiers,
                                       const bool keyDown)
+    {
+        auto keySentArgs = winrt::make<KeySentEventArgs>(vkey, scanCode, modifiers, keyDown);
+        _KeySentHandlers(*this, keySentArgs);
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:538]: Beware, this will cause a lot of needless IPC traffic.

Should reporting key events up to the app layer instead be the responsibility of `TermControl` (the last layer that knows about XAML?)

I know these are *not* xaml events, but handling it here has implications on the WPF control and eventually when Terminal is multi-process, will cause IPC amplification (hitting a key causes a synchronous (!) handler raise from Content Process -> Window Process, then a heap of synchronous (!) calls back down into `BroadcastKeyEvent` from Window Proc -> N Content Processes.)

Comment (DHowett) [reply src/cascadia/TerminalControl/ControlCore.cpp:538]: If I'm being completely honest, I don't think the control should be involved at all. Broadcasting is an application concern. It only "has" to be this way because we made key _bindings_ the control's concern.

Given that, though, we should restrict the damage across the layers as much as possible.

**Review by DHowett — CHANGES_REQUESTED** [review]
Ideology discussion

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 2 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalControl/TermControl.cpp]: I think I'm gonna go with `RawWriteKeyEvent`. `IDirectKeyListener` does need to broadcast up too, so I'm not sure we can just converge these 😕

File: src/cascadia/TerminalApp/AppActionHandlers.cpp [hunk 4]
```diff
@@ -1182,4 +1182,15 @@ namespace winrt::TerminalApp::implementation
...
     }
+    void TerminalPage::_HandleToggleBroadcastInput(const IInspectable& /*sender*/,
+                                                   const ActionEventArgs& args)
+    {
+        if (const auto activeTab{ _GetFocusedTabImpl() })
+        {
+            activeTab->ToggleBroadcastInput();
+        }
+
+        args.Handled(true);
```
Comment (DHowett) [inline src/cascadia/TerminalApp/AppActionHandlers.cpp:1193]: We may want to only mark this as handled if there was a terminal tab (consider pressing Ctrl+B or whatever in the settings tab -- it should go through)

File: src/cascadia/TerminalApp/App.xaml [hunk 5]
```diff
@@ -193,6 +199,9 @@
 
                             <StaticResource x:Key="SettingsUiTabBrush"
                                             ResourceKey="SystemControlBackgroundBaseLowBrush" />
+
+                            <SolidColorBrush x:Key="BroadcastPaneBorderColor"
+                                             Color="{StaticResource SystemAccentColorLight2}" />
```
Comment (DHowett) [inline src/cascadia/TerminalApp/App.xaml:204]: Using this in HC may cause trouble -- consider selecting a different color

File: src/cascadia/TerminalApp/Pane.cpp [hunk 6]
```diff
@@ -3159,3 +3178,76 @@ void Pane::CollectTaskbarStates(std::vector<winrt::TerminalApp::TaskbarState>& s
...
+        if (_control != sourceControl && !_control.ReadOnly())
+        {
+            _control.BroadcastChar(character, scanCode, modifiers);
+        }
+    }
+    else
+    {
+        _firstChild->BroadcastChar(sourceControl, character, scanCode, modifiers);
+        _secondChild->BroadcastChar(sourceControl, character, scanCode, modifiers);
+    }
```
Comment (DHowett) [inline src/cascadia/TerminalApp/Pane.cpp]: nit: we've been moving away from manual tree walks and towards `WalkTree<T>`. We should do that for all of these as well.

File: src/cascadia/TerminalControl/ControlCore.cpp [hunk 7]
```diff
@@ -399,6 +399,8 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         _sendInputToConnection(wstr);
     }
 
+    // Method Description:
+    // - Broadcasts this char to any registered CharSent handlers
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.cpp:403]: no longer true; no need for the diff this way!

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 8]
```diff
@@ -1204,6 +1217,18 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
                                        const bool keyDown)
+    {
+        // Broadcast the key  to all listeners
+        auto keySentArgs = winrt::make<KeySentEventArgs>(vkey, scanCode, modifiers, keyDown);
+        _KeySentHandlers(*this, keySentArgs);
+
+        return BroadcastKeyEvent(vkey, scanCode, modifiers, keyDown);
+    }
+
+    bool TermControl::BroadcastKeyEvent(const WORD vkey,
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp]: This isn't exactly true, is it? It feels weird for the final port-of-call before we get out of TermControl to be called "Broadcast". To whom does it broadcast?

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 9]
```diff
@@ -948,10 +948,23 @@ namespace winrt::Microsoft::Terminal::Control::implementation
             modifiers |= ControlKeyStates::EnhancedKey;
         }
 
-        const auto handled = _core.SendCharEvent(ch, scanCode, modifiers);
+        // Broadcast the character to all listeners
+        auto charSentArgs = winrt::make<CharSentEventArgs>(ch, scanCode, modifiers);
+        _CharSentHandlers(*this, charSentArgs);
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp:953]: one last thing I'd be way of is the object cost. Every character, every key, is getting another heap allocated object that is bubbled up to somebody who is listening. It's not likely to be a bottleneck -- human typing speed is limited, after all -- but it's worth considering.

File: src/cascadia/TerminalControl/ControlCore.h [hunk 10]
```diff
@@ -232,6 +232,8 @@ namespace winrt::Microsoft::Terminal::Control::implementation
         TYPED_EVENT(ShowWindowChanged,         IInspectable, Control::ShowWindowArgs);
         TYPED_EVENT(UpdateSelectionMarkers,    IInspectable, Control::UpdateSelectionMarkersEventArgs);
         TYPED_EVENT(OpenHyperlink,             IInspectable, Control::OpenHyperlinkEventArgs);
+        TYPED_EVENT(KeySent,                   IInspectable, Control::KeySentEventArgs);
```
Comment (DHowett) [inline src/cascadia/TerminalControl/ControlCore.h:235]: Is this part of ControlCore still?

**Review by DHowett — CHANGES_REQUESTED** [review]
Okay, now that I've nitted more naming concerns (with apologies; I just think that while we're doing this we might as well get it right...) I've got some deeper worries!

Does anything that inputs a whole string at once work? This includes...
- IME
- Clipboard
- Drag/drop a filename

If I had to guess, the one most likely to cause folks trouble is the clipboard. Rarely am I going to key in a long command that I wanted to send to four servers. I'm almost certainly gonna paste it.

Also, what does mouse input do? *Should* it do something? I imagine clicking on things that are different sizes and locations will just suck... but what about using the scroll wheel?

I guess this is where the layering discussion comes up again. Bah.

**Comment by zadjii-msft** [comment]
WIP, it's the EOD here

* [ ] IME
* [ ] Clipboard
  - `TerminalPage::_PasteText` this should be easy enough to check if broadcast is available and paste to all the controls. 
* [ ] Drag/drop a filename

File: src/cascadia/TerminalApp/App.xaml (hunk 5 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/App.xaml:204]: picked a new one and ngl I love it
image

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 11]
```diff
@@ -2423,6 +2423,24 @@ namespace winrt::TerminalApp::implementation
...
+        // First, check if we're in broadcast input mode. If so, let's tell all
+        // the controls to paste.
+        if (const auto& tab{ _GetFocusedTabImpl() })
+        {
+            if (tab->TabStatus().IsInputBroadcastActive())
+            {
+                tab->GetRootPane()->WalkTree([](auto&& pane) {
+                    if (auto control = pane->GetTerminalControl())
+                    {
+                        control.PasteTextFromClipboard();
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalPage.cpp:2435]: Idle thought... should this not go via the same path as RawWriteString? PasteTextFromClipboard means "ask the app for the clipboard content," which could result in one ContentDialog per control (!) confirming the paste.

I'd expect this:

```mermaid
sequenceDiagram
    TerminalPage-->>TermControl: PasteTextFromClipboard
    TermControl-->>TerminalPage: GetTextFromClipboard
    TerminalPage-->>TermControl: Here's the text bud
    TermControl-->>Pane: SentString("here's the...")
    Pane-->>OtherTermControl: RawWriteString("here's the...")
    Pane-->>OtherTermControl: RawWriteString("here's the...")
    Pane-->>OtherTermControl: RawWriteString("here's the...")
```

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2435]: note from discussion: bracketed paste, make sure that works

File: src/cascadia/TerminalApp/Pane.cpp [hunk 12]
```diff
@@ -3159,3 +3178,75 @@ void Pane::CollectTaskbarStates(std::vector<winrt::TerminalApp::TaskbarState>& s
...
+{
+    if (_lastActive)
+    {
+        return s_focusedBorderBrush;
+    }
+
+    if (_broadcastEnabled)
+    {
+        return s_broadcastBorderBrush;
+    }
```
Comment (PankajBhojwani) [inline src/cascadia/TerminalApp/Pane.cpp:3249]: Should we not set the color to `s_broadcastBorderBrush` if the pane is in read only mode? It might be confusing since the broadcasted input won't be sent to it but it has the same border colors as the others.

File: src/cascadia/TerminalApp/TerminalTab.cpp [hunk 13]
```diff
@@ -1629,4 +1673,12 @@ namespace winrt::TerminalApp::implementation
 
         return Title();
     }
+
+    // Method Description:
+    // - Toggle read-only mode on the active pane
```
Comment (PankajBhojwani) [inline src/cascadia/TerminalApp/TerminalTab.cpp:1678]: Nit: method description needs to be updated

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 14]
```diff
@@ -948,10 +950,28 @@ namespace winrt::Microsoft::Terminal::Control::implementation
...
+
         e.Handled(handled);
     }
 
+    bool TermControl::RawWriteChar(const wchar_t character,
+                                   const WORD scanCode,
+                                   const winrt::Microsoft::Terminal::Core::ControlKeyStates modifiers)
+    {
+        const auto handled = _core.SendCharEvent(character, scanCode, modifiers);
+        return handled;
```
Comment (PankajBhojwani) [inline src/cascadia/TerminalControl/TermControl.cpp:967]: nit:
```suggestion
        return _core.SendCharEvent(character, scanCode, modifiers);
```

**Review by PankajBhojwani — CHANGES_REQUESTED** [review]
Just concerned about the border color in read only mode, rest are nits

**Review by PankajBhojwani — APPROVED** [review]

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 11 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2587]: Hmm. This is... less trivial. 

`TerminalPage::_PasteFromClipboardHandler` calls the event's `HandleClipboardData`, which is a callback that `Interactivity` set up to call `_core->PasteText` [^1] when the paste is completed. However, Core & Interactivity don't actually know anything about broadcast mode. That's all good and well.

I suppose this could be done by finagling the paste handling a bit. We'd bubble up to TermControl, it'd register the paste callback itself, and in its own handler it could decide to re-raise a `event Windows.Foundation.TypedEventHandler<Object, StringSentEventArgs> PasteSent;`[^2], which pane could listen to, and then propagate. Feels kinda extra dirty, tbh.

I don't think the dialog will pop multiple times, on account of the fact that there can only ever be one dialog. I'll have to validate that though. 

Alternatively, we could do the broadcast check inside `TerminalPage::_PasteFromClipboardHandler` and just call `WalkTree([text](){ p->Control().PasteText(); }); and plumb that through. ~Might be tricky to determine that though. `TermControl` doesn't expose the `Interactivity` out of it, so we couldn't compare the `sender`~ well, I suppose if we're not in broadcast mode, we could just call the event callback as before. 

Thoughts?


[^1]: `core->PasteText` is what handles bracketed paste, so that's fine.
[^2]: can't be the StringSent, cause we do need to special-case pastes

Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2738]: @DHowett gibe thoughts pls, so I can merge this when I get back from break for 1.19 selfhost ☺️

**Comment by zadjii-msft** [comment]
Notes from Teams, abridged

> so, the thing that it hinges on for me is
> If you have two panes selected, and one is in `DECKPAM` and the other is in `DECCKM` or whatever... does one arrow key send two _different things_ properly, or does it send the same thing to both panes?
> so, the thing is
> if we choose to send the same content on either side, I think that makes our lives easier... and talking with Leonard I am much more comfortable with the "TerminalControl raises an event when it sends text" solution
> with one caveat
> _that event must not ever be raised from any layer_  unless someone is listening
> the cool thing is, events can be checked for truthiness - we don't need to add a separate boolean to plumb down to ControlCore
> Leonard talked me into this if i recall
> if the answer is 1 ("they send different things") we need to bifurcate all event generation
> which.. ugh

> okay so
> tldr: make sure we don't raise events unless there's a listener

> yea... and we can kill the bifurcated "paste" and everything else handlers
> and boil it down to the simple "all terminal input sent via Write gets reflected"
> like your OG work IIRC

> I have been working on reviving the wpf m.t.c.dll
> and looking at the public API of TerminalControl
> and it's sorta all over the place admittedly but
> this feels less intrusive for the API than having to poke through all the different types of user input as event callers
> you know?

> as usual, you come up with the right solution to begin with and I say something dumb like "can it be totally different" before coming all the way back to a modified version of your original idea

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 11 above)
Comment (DHowett) [reply src/cascadia/TerminalApp/TerminalPage.cpp:2799]: We chatted last-ish week about RawWriteString taking a "StringKind" of Clipboard, Raw, Whatever. Noting that here to close the loop. :)

File: src/cascadia/TerminalApp/TerminalTab.cpp [hunk 15]
```diff
@@ -1763,4 +1769,88 @@ namespace winrt::TerminalApp::implementation
...
     }
+
+    // Method Description:
+    // - Toggle broadcasting input to all the panes in this tab.
+    void TerminalTab::ToggleBroadcastInput()
+    {
+        _tabStatus.IsInputBroadcastActive(!_tabStatus.IsInputBroadcastActive());
+
+        const bool isBroadcasting = _tabStatus.IsInputBroadcastActive();
+        _rootPane->EnableBroadcast(isBroadcasting);
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalTab.cpp:1780]: nit: fold or reverse?

```suggestion
        const bool newIsBroadcasting = !_tabStatus.IsInputBroadcastActive();
        _tabStatus.IsInputBroadcastActive(newIsBroadcasting);
        _rootPane->EnableBroadcast(newIsBroadcasting);
```

("Set on X and then ask X" weirds me out)

File: src/cascadia/TerminalApp/TerminalTab.cpp [hunk 16]
```diff
@@ -1763,4 +1769,88 @@ namespace winrt::TerminalApp::implementation
...
+                                }
+                            }
+                        });
+                    }
+
+                    else
+                    {
+                        // REVOKE EVENT HANDLERS
+
+                        control.KeySent(events.keySentToken);
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalTab.cpp:1848]: Do we need to clear out the values here in case we end up in a torn state?

File: src/cascadia/TerminalApp/TerminalTab.cpp [hunk 17]
```diff
@@ -1763,4 +1769,88 @@ namespace winrt::TerminalApp::implementation
...
+                        control.CharSent(events.charSentToken);
+                        control.StringSent(events.stringSentToken);
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalTab.cpp:1850]: dumb question: do we properly revoke when you move pane to tab (!) or move tab to window (?)

File: src/cascadia/TerminalApp/TerminalTab.cpp [hunk 18]
```diff
@@ -1763,4 +1769,88 @@ namespace winrt::TerminalApp::implementation
...
+                return;
+            }
+            if (const auto& control{ p->GetTerminalControl() })
+            {
+                auto it = _controlEvents.find(*paneId);
+                if (it != _controlEvents.end())
+                {
+                    auto& events = it->second;
+
+                    if (isBroadcasting)
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalTab.cpp:1801]: If you take the above,

```suggestion
                    if (newIsBroadcasting)
```

File: src/cascadia/TerminalApp/TerminalTab.cpp [hunk 19]
```diff
@@ -1763,4 +1769,88 @@ namespace winrt::TerminalApp::implementation
...
+                    {
+                        // ADD EVENT HANDLERS HERE
+                        events.keySentToken = control.KeySent([weakThis](auto&& sender, auto&& e) {
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalTab.cpp:1804]: in theory if we use auto revokers, we could simplify the code to this:

```
events.keySentToken.revoke();
xxx.revoke();

if (newIsBroadcasting) {
    keySentToken = foo;
    xxx = bar;
}
```

and simplify the else case plus stomp the event handlers properly every time

File: src/cascadia/TerminalApp/TerminalTab.cpp [hunk 20]
```diff
@@ -942,7 +948,7 @@ namespace winrt::TerminalApp::implementation
             }
         });
 
-        events.readOnlyToken = control.ReadOnlyChanged([dispatcher, weakThis](auto&&, auto&&) -> winrt::fire_and_forget {
+        events.readOnlyToken = control.ReadOnlyChanged([dispatcher, weakThis](auto&&, auto &&) -> winrt::fire_and_forget {
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalTab.cpp:951]: weird, code formatter did us dirty

File: src/cascadia/TerminalApp/TerminalTab.cpp [hunk 21]
```diff
@@ -620,7 +622,7 @@ namespace winrt::TerminalApp::implementation
                 _nextPaneId++;
             }
         });
-
+        pane->EnableBroadcast(_tabStatus.IsInputBroadcastActive());
```
Comment (DHowett) [inline src/cascadia/TerminalApp/TerminalTab.cpp:625]: this is the scary for me - will a pane moved into a different tab continue broadcasting to the old tab? we don't do any event hookups or unhooks here.

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 22]
```diff
@@ -2587,7 +2630,8 @@ namespace winrt::Microsoft::Terminal::Control::implementation
             return;
         }
 
-        _core.SendInput(text);
+        // SendInput will take care fo broadcasting for us.
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp:2633]:
```suggestion
        // SendInput will take care of broadcasting for us.
```

File: src/cascadia/TerminalControl/TermControl.cpp [hunk 23]
```diff
@@ -2812,11 +2856,21 @@ namespace winrt::Microsoft::Terminal::Control::implementation
                     allPathsString += fullPath;
                 }
 
-                _core.PasteText(winrt::hstring{ allPathsString });
+                _pasteTextWithBroadcast(winrt::hstring{ allPathsString });
             }
         }
     }
 
+    void TermControl::_pasteTextWithBroadcast(const winrt::hstring& text)
```
Comment (DHowett) [inline src/cascadia/TerminalControl/TermControl.cpp:2864]: This name really confused me - it actually ISN'T used in paste(!) but it made me double-take because we are manually broadcasting the paste event and the incoming text doesn't come back through here. Does it? Or, does it...

**Review by DHowett — CHANGES_REQUESTED** [review]

File: src/cascadia/TerminalApp/TerminalTab.cpp (hunk 21 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalTab.cpp:625]: Nope, works like a charm

File: src/cascadia/TerminalApp/TerminalTab.cpp (hunk 17 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalTab.cpp:1850]: Sure seem to

File: src/cascadia/TerminalApp/Pane.cpp [hunk 24]
```diff
@@ -3177,3 +3179,76 @@ void Pane::CollectTaskbarStates(std::vector<winrt::TerminalApp::TaskbarState>& s
...
+}
+
+void Pane::BroadcastKey(const winrt::Microsoft::Terminal::Control::TermControl& sourceControl,
+                        const WORD vkey,
+                        const WORD scanCode,
+                        const winrt::Microsoft::Terminal::Core::ControlKeyStates modifiers,
+                        const bool keyDown)
+{
+    WalkTree([&](auto pane) {
```
Comment (lhecker) [inline src/cascadia/TerminalApp/Pane.cpp:3203]: nit: Unnecessary copy by value. (2x below as well.)

File: src/cascadia/TerminalApp/TerminalPage.cpp [hunk 25]
```diff
@@ -4570,6 +4588,21 @@ namespace winrt::TerminalApp::implementation
             // will eat focus.
             _paneResources.unfocusedBorderBrush = SolidColorBrush{ Colors::Black() };
         }
+
+        const auto broadcastColorKey = winrt::box_value(L"BroadcastPaneBorderColor");
+        if (res.HasKey(broadcastColorKey))
+        {
+            // MAKE SURE TO USE ThemeLookup
```
Comment (lhecker) [inline src/cascadia/TerminalApp/TerminalPage.cpp:4595]: Make sure to... why? 😅

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 23 above)
Comment (lhecker) [inline src/cascadia/TerminalControl/TermControl.cpp:2864]: FYI technically you can use `winrt::param::hstring` here for 0 copy string passing, but I've heard it's considered an internal API so we can't use it. 😟 (Or can we? Technically we are in control over our code... Oh well...)

**Review by lhecker — APPROVED** [review]
I really don't like how the control flow goes from UI to TermControl, then back up into UI, then back down into TermControl. That seems kind of unnecessary tbh. 😅 Why can't it distribute the input across all TermControls from the get-go?
But since I'm wholly unqualified to say how I'd actually do it differently (maybe this is the only way to do it?), I can't see it as an actual fault in the PR either. ✅

File: src/cascadia/TerminalApp/TerminalPage.cpp (hunk 25 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalPage.cpp:4595]: Oh that comment is explained at length like, two code blocks up

>             // MAKE SURE TO USE ThemeLookup, so that we get the correct resource for
>             // the requestedTheme, not just the value from the resources (which
>             // might not respect the settings' requested theme)

File: src/cascadia/TerminalApp/TerminalTab.cpp (hunk 16 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalTab.cpp]: resolving on account of switching to revokers

File: src/cascadia/TerminalApp/TerminalTab.cpp (hunk 21 above)
Comment (DHowett) [reply src/cascadia/TerminalApp/TerminalTab.cpp:625]: Huh! Can you walk me through why it works?
As a reviewer I couldn't see the machinery that kept it safe. I'd rather know that we know the code is safe than that we tested and found the behavior safe :smile:

File: src/cascadia/TerminalControl/TermControl.cpp (hunk 23 above)
Comment (DHowett) [reply src/cascadia/TerminalControl/TermControl.cpp:2864]: (Since I was re-requested for review) _Does_ this have anything to do with pasting? It looks like it doesn't.

**Comment by DHowett** [comment]
> Why can't it distribute the input across all TermControls from the get-go?

Four words: XAML

**Review by DHowett — APPROVED** [review]
I trust you that it works for moving panes around, but I would love to know why. Approved with comments -- that "paste" thing still confuses me :)

File: src/cascadia/TerminalApp/TerminalTab.cpp (hunk 21 above)
Comment (zadjii-msft) [reply src/cascadia/TerminalApp/TerminalTab.cpp:625]: Okay, I know this is gonna cause you physical pain to read:


* `TerminalPage::_HandleMovePane`
  * `TerminalPage::_MovePane`
    * `TerminalTab::DetachPane`
      * `Pane::DetachPane`
        * `this._DetachedHandlers` -> handled in a lambda in `TerminalTab::_AttachEventHandlersToPane`
          * revokes all our per-pane event handlers
          * `TerminalTab::_DetachEventHandlersFromControl`
            * revokes per-control handlers (including these)
    * `TerminalTab::AttachPane`
      * reattaches them to the new tab.

**Comment by zadjii-msft** [comment]
leaving here for future me 
broadcast-input-demo-000
